contracts/extensions/contracts/BalanceThresholdFilter/MixinBalanceThresholdFilterCore.sol

```diff
@@ -0,0 +1,323 @@
+            signature
+        );
+    }
+
+    /// @dev Validates addresses meet the balance threshold specified by `BALANCE_THRESHOLD`
+    ///      for the asset `THRESHOLD_ASSET`. If one address does not meet the thresold
+    ///      then this function will revert. Which addresses are validated depends on
+    ///      which Exchange function is to be called (defined by `signedExchangeTransaction` above).
+    ///      No parameters are taken as this function reads arguments directly from calldata, to save gas.
+    ///      If all addresses are valid then this function emits a ValidatedAddresses event, listing all
+    ///      of the addresses whose balance thresholds it checked.
+    function validateBalanceThresholdsOrRevert()
+        internal
+    {
+        // Addresses that are validated below.
+        address[] memory validatedAddresses;
+
+        ///// Do not add variables after this point.         /////
+        ///// The assembly block may overwrite their values. /////
+
+        // Validate addresses
+        assembly {
+             /// @dev Emulates the `calldataload` opcode on the embedded Exchange calldata,
+             ///      which is accessed through `signedExchangeTransaction`.
+             /// @param offset - Offset into the Exchange calldata.
+             /// @return value - Corresponding 32 byte value stored at `offset`.
+            function exchangeCalldataload(offset) -> value {
+                // Pointer to exchange transaction
+                // 0x04 for calldata selector
+                // 0x40 to access `signedExchangeTransaction`, which is the third parameter
+                let exchangeTxPtr := calldataload(0x44)
+
+                // Offset into Exchange calldata
+                // We compute this by adding 0x24 to the `exchangeTxPtr` computed above.
+                // 0x04 for calldata selector
+                // 0x20 for length field of `signedExchangeTransaction`
+                let exchangeCalldataOffset := add(exchangeTxPtr, add(0x24, offset))
+                value := calldataload(exchangeCalldataOffset)
+            }
+
+            /// @dev Convenience function that skips the 4 byte selector when loading
+            ///      from the embedded Exchange calldata.
+            /// @param offset - Offset into the Exchange calldata (minus the 4 byte selector)
+            /// @return value - Corresponding 32 byte value stored at `offset` + 4.
+            function loadExchangeData(offset) -> value {
+                value := exchangeCalldataload(add(offset, 0x4))
+            }
+
+             /// @dev A running list is maintained of addresses to validate. 
+             ///     This function records an address in this array.
+             /// @param addressToValidate - Address to record for validation.
+            function recordAddressToValidate(addressToValidate) {
+                // Compute `addressesToValidate` memory offset
+                let addressesToValidate_ := mload(0x40)
```

Comment:
It might be cleaner to do this without creating new temporary variables (I find these variable names confusing). All we really need to do is 1. increment length, 2. store new address at free memory ptr, 3. increment free memory ptr.